handle t.Parallel in fixtures

IlyasYOY opened the pull request:
Closes: https://github.com/kunwardeep/paralleltest/issues/56

IlyasYOY committed:
feat(testdata): add test cases for parallel testing with helpers

Add new test cases to verify the behavior of parallel tests with and without the use of `t.Helper()`. This includes:
- `TestUsingHelper`: Uses `t.Helper()` and `t.Parallel()`.
- `TestUsingHelperWithoutHelper`: Does not use `t.Helper()` but uses `t.Parallel()`.
- `TestUsingHelperWithoutParallel`: Uses `t.Helper()` but does not use `t.Parallel()`, expecting a linting warning.

For: https://github.com/kunwardeep/paralleltest/issues/56

IlyasYOY committed:
refactor(paralleltest): enhance test function analysis to support helpers with t.Parallel

Enhance the `parallelAnalyzer` to correctly identify and analyze test functions that utilize unexported helper functions with `t.Parallel`. Introduce a new `funcInfo` struct to store function declarations and their corresponding files. Modify the `run` method to collect all function declarations from test files and store them in `funcDecls`. Update the `analyzeTestFunction` method to inspect function bodies for calls to unexported helper functions and check if they contain `t.Parallel`. Add a new `hasParallelInHelpers` method to recursively check if a function or its unexported helpers call `t.Parallel`. Include test cases in `cross_file_helper_test.go` and `t_test.go` to verify the functionality.

Closes: https://github.com/kunwardeep/paralleltest/issues/56

IlyasYOY commented on pkg/paralleltest/paralleltest.go:
Changes here are mainly generated, I went through and think they lgtm. 

The PR has tests, this is good.

Thanks you for the `analysistest` usage, it makes writing tests a breeze. I simply wrote the test, AI came up with the solution.

IlyasYOY committed:
refactor(paralleltest): improve readability of parallel check in helpers

Refactor the `hasParallelInHelpers` function to improve readability by simplifying the nested conditionals and using early returns. This makes the logic easier to follow and understand.

IlyasYOY committed:
fix(paralleltest): add mutex to ensure thread safety in run method

Add a mutex to the `parallelAnalyzer` struct and lock it in the `run` method to ensure thread safety when collecting function declarations from test files. This prevents potential race conditions when multiple goroutines access the `funcDecls` map concurrently.

IlyasYOY committed:
refactor(paralleltest): remove mutex and funcDecls from parallelAnalyzer

Remove the `mutex` and `funcDecls` fields from the `parallelAnalyzer` struct and update the `analyzeTestFunction` and `hasParallelInHelpers` methods to accept `funcDecls` as a parameter. This change simplifies the struct and ensures that the function declarations are passed explicitly, avoiding potential concurrency issues and making the code easier to understand and maintain.

IlyasYOY commented:
Hello @kunwardeep! Thank you for running the checks on the PR. 

Maybe you have any question on the implementation of the feature? I'd be glad to answer them and deliver fixes to the issues you might find.

I guess I have to hide the feature behind the flag? I think it'd be nice to have! The reason I did not add this in the first place: I think the change is backward compatible. But (there is always a but) there might be problems for users that use `nolintlint` [linter](https://github.com/ashanbrown/nolintlint) from golangci-lint in `-specific` mode. Linter might fail because there won't be any `paralleltest` violations.

Thanks!

kunwardeep commented:
Hey mate

Thanks for the PR. I was just a bit busy with work. I'll try and get to it
this week

Cheers
KD

On Sat, Nov 29, 2025 at 8:51 PM Ilya Ilyinykh ***@***.***>
wrote:

> *IlyasYOY* left a comment (kunwardeep/paralleltest#57)
> <https://github.com/kunwardeep/paralleltest/pull/57#issuecomment-3591517793>
>
> Hello @kunwardeep
> <https://github.com/kunwardeep>!
> Thank you for running the checks on the PR.
>
> Maybe you have any question on the implementation of the feature? I'd be
> glad to answer them and deliver fixes to the issues you might find.
>
> I guess I have to hide the feature behind the flag? I think it'd be nice
> to have! The reason I did not add this in the first place: I think the
> change is backward compatible. But (there is always a but) there might be
> problems for users that use nolintlint linter
> <https://github.com/ashanbrown/nolintlint>
> from golangci-lint in -specific mode. Linter might fail because there
> won't be any paralleltest violations.
>
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kunwardeep/paralleltest/pull/57#issuecomment-3591517793>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADEZLTQWONX2EZAMC3L7DRT37FUA5AVCNFSM6AAAAACLOUCNUWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOJRGUYTONZZGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Disclaimer

This email and any files transmitted with it are confidential and intended solely for the individual or entity to whom they are addressed. Please notify the sender immediately by e-mail if you have received this e-mail by mistake and delete this e-mail from your system. You may neither use nor edit any such emails including attachments, nor make them accessible to third parties in any manner whatsoever. Easygo makes no representation as to the accuracy of information contained in this email (and any attachments), including historical or forward-looking statements. Easygo and its related bodies corporate accept no liability for any damage caused by this email or its attachments due to viruses, interference, interception or unauthorised access.

IlyasYOY commented:
@kunwardeep Hey! I know you might be busy at the moment.

I'm asking if you could take a look at mine (or https://github.com/kunwardeep/paralleltest/pull/58). I know a lot of people use your linter and it'd be good for them to benefit and get support for `t.Helper` (or both `t.Helper`, `t.Chdir`, `t.Setenv`). Especially considering the monolith-architecture of golangci-lint (you cannot easily add your own fork in there).

I don't want to haste you, just a reminder, maybe these PR got lost.

IlyasYOY commented on pkg/paralleltest/paralleltest.go:
```diff
@@ -497,3 +551,57 @@ func loopVarReferencedInRun(call *ast.CallExpr, vars []types.Object, typeInfo *t
+		return false
+	}
+	visited[funcDecl.Name.Name] = true
+	defer func() { delete(visited, funcDecl.Name.Name) }()
```
I think the delete might be left, because it doesn't violate the loop detection logic. 

Items will be removed at the ascending of the recursion, so items causing cycles will be removed after the cycle was detected.

I understand the concern so I added test for recursive helper calls.

IlyasYOY committed:
docs(paralleltest): improve comments and clarify recursion handling

Fix typo in comment describing test context parameter and add explanation for visited map usage to prevent infinite recursion in helper analysis.

IlyasYOY committed:
test(paralleltest): add recursive helper tests for parallel detection

Add tests that verify detection of missing t.Parallel calls in recursive helper functions, covering both without and with parallel scenarios.
